### DISCUSSION TITLE
Custom github regex updater not finding updates

### DISCUSSION THREAD

**Discussion by DanySK** [renovatebot/renovate#17342] [Request Help] [answered]
Hi,
I wrote a custom regex updater to update images in markdown files imported from other repos.

The config is:

```json
  "regexManagers": [
      {
        "fileMatch": ["^.*\\.md$"],
        "matchStrings": [
            "(?<comment><!--\\s*from-branch:\\s*(?<branch>[^\\s]+)\\s*-->)?\\s*!\\[(?<imageName>.*)\\]\\(https:\\/\\/raw\\.githubusercontent\\.com\\/(?<slug>(?<owner>[^\\/]+)\\/(?<repo>[^\\/]+))\\/(?<currentDigest>[^\\/]+)\\/(?<path>.+)\\)"
        ],
        "depNameTemplate": "image {{#if imageName}}{{{imageName}}}{{else}}{{{slug}}}/{{{path}}}{{/if}}",
        "currentValueTemplate": "{{#if branch}}{{{branch}}}{{else}}images{{/if}}",
        "packageNameTemplate": "https://github.com/{{{slug}}}.git",
        "datasourceTemplate": "git-refs",
        "autoReplaceStringTemplate": "{{{comment}}}\n[{{{imageName}}}](https://raw.githubusercontent.com/{{{slug}}}/{{{newDigest}}}/{{{path}}})",
        "versioningTemplate": "git"
      }
  ]
```

in short, it tries to match markdown images from raw.githubusercontent.com, and then use the git data source to fetch updates from GitHub, defaulting to the `images` branch unless a comment configures the system to do otherwise.

I'm exercising this system on a repo containing a markdown file with the following string:

```markdown
![silos](https://raw.githubusercontent.com/DanySK/shared-slides/14156e4740992d1fc1ffc62c53aaae2bc21c2729/devops/silos.svg)
```

I would expect Renovate to propose an update to:
```markdown
![silos](https://raw.githubusercontent.com/DanySK/shared-slides/62522124c1ee8be3d463291707b974ff32518d78/devops/silos.svg)
```
as there is a newer commit on the `images` branch.

Renovate seems to run the match just fine, but finds no updates:

```json
{
  "baseBranch": "master",
  "config": {
    "regex": [
      {
        "packageFile": "devops/devops-intro.md",
        "deps": [
          {
            "depName": "image silos",
            "packageName": "https://github.com/DanySK/shared-slides.git",
            "currentValue": "images",
            "currentDigest": "14156e4740992d1fc1ffc62c53aaae2bc21c2729",
            "datasource": "git-refs",
            "versioning": "git",
            "replaceString": "\n\n![silos](https://raw.githubusercontent.com/DanySK/shared-slides/14156e4740992d1fc1ffc62c53aaae2bc21c2729/devops/silos.svg)",
            "depIndex": 0,
            "warnings": [],
            "updates": []
          }
        ],
        "matchStrings": [
          "(?<comment><!--\\s*from-branch:\\s*(?<branch>[^\\s]+)\\s*-->)?\\s*!\\[(?<imageName>.*)\\]\\(https:\\/\\/raw\\.githubusercontent\\.com\\/(?<slug>(?<owner>[^\\/]+)\\/(?<repo>[^\\/]+))\\/(?<currentDigest>[^\\/]+)\\/(?<path>.+)\\)"
        ],
        "depNameTemplate": "image {{#if imageName}}{{{imageName}}}{{else}}{{{slug}}}/{{{path}}}{{/if}}",
        "packageNameTemplate": "https://github.com/{{{slug}}}.git",
        "currentValueTemplate": "{{#if branch}}{{{branch}}}{{else}}images{{/if}}",
        "datasourceTemplate": "git-refs",
        "versioningTemplate": "git",
        "autoReplaceStringTemplate": "{{{comment}}}\n[{{{imageName}}}](https://raw.githubusercontent.com/{{{slug}}}/{{{newDigest}}}/{{{path}}})"
      }
    ]
  }
}
```

Any idea of what I am getting wrong?
Thanks!

**Comment by viceice** [comment]
current value seems to be invalid

**Comment by DanySK** [comment, marked as answer]
Hi @viceice,
shouldn't it match the branch name?
It is a valid branch name: https://github.com/DanySK/shared-slides/tree/images.

Since I wanted to mimic the behavior of submodules updates, I tried to build a similar update object. This is a working submodules update:

```json
{
  "baseBranch": "master",
  "config": {
    "git-submodules": [
      {
        "packageFile": ".gitmodules",
        "deps": [
          {
            "depName": "slides/content/extra/pps-ci-introduction/PPS-ci-examples",
            "packageName": "https://github.com/APICe-at-DISI/PPS-ci-examples.git",
            "currentValue": "master",
            "currentDigest": "4a1f2ae14d16afd08da462c98b99a38b536111e0",
            "depIndex": 0,
            "updates": [
              {
                "updateType": "digest",
                "newValue": "master",
                "newDigest": "8d14751065a2f524e51c2df24e5eba302c004700",
                "branchName": "renovate/slides-content-extra-pps-ci-introduction-pps-ci-examples-digest"
              }
            ],
            "warnings": [],
            "versioning": "git"
          }
        ],
        "datasource": "git-refs"
      }
    ]
  }
}
```

apparently `currentValue` gets bound to the branch name...

**Reply by viceice** [reply]
you can't use arbitrary regex groups. only known are allowed. you probably need to limit that.

maybe you should also remove the auto replace template

**Reply by DanySK** [reply]
@viceice where do I find the allowed ones?
Are you sure that this limitation is in place? Look at `"depName": "image silos"`: it is clearly reusing the value of capture group `imageName`, even though it is not in the list of the allowed ones (or is it? It'd be a pretty lucky shot to match a supported name like this).

 How do I tell Renovate how to replace the string but with `autoReplaceTemplate`? Did I misunderstand the goal of the field?

The doc is pretty patchy for the regex manager. Are there more advanced examples than those in the docs?

**Reply by DanySK** [reply]
@viceice it also seems to be matching and reusing the group named `slug` to populate `packageName` via the corresponding template... Is there a way to have more debug information on the update lookup?

**Reply by DanySK** [reply]
@viceice looks like removing `autoRemplaceTemplate` did the trick, indeed: https://github.com/DanySK/shared-slides/pull/2/files :partying_face: 

What is the field meant for?

**Reply by viceice** [reply]
this are the allowed match group names:
https://github.com/renovatebot/renovate/blob/229643f7e426192ba1bcb53d5ad6cf11bf304341/lib/modules/manager/regex/utils.ts#L9-L19

**Reply by DanySK** [reply]
@viceice well, then maybe it is magic, but it is matching also `slug` and `imageName`. I guess (but I can't find the code) that at some point all regex groups get extracted, those that are named as fields get injected as such, and those that are not are still available when the templating engine is executed (which makes a lot of sense).

**Reply by viceice** [reply]
`autoReplaceStringTemplate` [^1] is used for special cases, it defines how renovate will create the replace value. You can only use defined renovate options, as it's compiled before replace, so the `matchStrings` groups are already gone.

[^1]: https://docs.renovatebot.com/configuration-options/#autoreplacestringtemplate

**Reply by viceice** [reply]
the `validMatchFields` are compiled on extraction with your match groups, so arbritrary regex groups work there. but `autoReplaceStringTemplate` is compiled at a later stage in autoreplace, so the match groups are no longer availabe.

https://github.com/renovatebot/renovate/blob/229643f7e426192ba1bcb53d5ad6cf11bf304341/lib/modules/manager/regex/utils.ts#L48-L64

**Reply by DanySK** [reply]
Oh, I see. I think that the use of custom groups would be very nice to document.
